daemon/hosts.go

```diff
@@ -0,0 +1,166 @@
+package daemon // import "github.com/docker/docker/daemon"
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	cerrdefs "github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/remotes/docker"
+	hostconfig "github.com/containerd/containerd/remotes/docker/config"
+	"github.com/docker/docker/registry"
+	"github.com/pkg/errors"
+)
+
+const (
+	defaultPath = "/v2"
+)
+
+// RegistryHosts returns the registry hosts configuration for the host component
+// of a distribution image reference.
+func (daemon *Daemon) RegistryHosts(host string) ([]docker.RegistryHost, error) {
+	hosts, err := hostconfig.ConfigureHosts(context.Background(), hostconfig.HostOptions{
+		// TODO: Also check containerd path when updating containerd to use multiple host directories
+		HostDir: hostconfig.HostDirFromRoot(registry.CertsDir()),
+	})(host)
+	if err == nil {
+		// Merge in legacy configuration if provided
+		if cfg := daemon.Config(); len(cfg.Mirrors) > 0 || len(cfg.InsecureRegistries) > 0 {
+			hosts, err = daemon.mergeLegacyConfig(host, hosts)
+		}
+	}
+
+	return hosts, err
+}
+
+func (daemon *Daemon) mergeLegacyConfig(host string, hosts []docker.RegistryHost) ([]docker.RegistryHost, error) {
+	// If no hosts provided, nothing to do.
+	// If multiple hosts provided, then a mirror configuration is already provided and
+	// should not overwrite with legacy config.
+	if len(hosts) == 0 || len(hosts) > 1 {
+		return hosts, nil
+	}
+	sc := daemon.registryService.ServiceConfig()
+	if host == "docker.io" && len(sc.Mirrors) > 0 {
+		var mirrorHosts []docker.RegistryHost
+		for _, mirror := range sc.Mirrors {
+			h := hosts[0]
+			h.Capabilities = docker.HostCapabilityPull | docker.HostCapabilityResolve
+
+			u, err := url.Parse(mirror)
+			if err != nil || u.Host == "" {
+				u, err = url.Parse(fmt.Sprintf("//%s", mirror))
```

Comment:
should this one use `//:` (with a colon)? if the intent is to make sure if has a scheme.

Also for this, we may as well skip the Sprintf, and just concat the strings

Reply:
This is what buildkit was doing but I agree it is not the approach done elsewhere. In containerd normally we parse as `dummy://%s` to extract a host and path.

The original code is https://github.com/moby/buildkit/blob/master/util/resolver/utils.go but it could use more unit testing.

Reply:
oh! wrote the colon in the wrong place 🙈

Hmm, actually maybe it was correct; scheme-less URL may not need the colon?

Reply:
The original logic was throwing away the scheme however that does not match how the mirror string was originally be used. Updated to parse with "dummy" scheme and override the scheme when provided. As it is now it will always use https even when explicitly setting the scheme. Also added a unit tests for this section so we can just add test cases to ensure it behaves as expected.